### PR TITLE
feat(raft): introduce read index

### PR DESCRIPTION
**Pull request by noneback**
Signed-off-by: noneback <chenlan_of_learn@126.com>

### CODE REVIEW

**Comment by noneback** [comment]
Draft. I only change the enforce-related code to see if it fits.

File: pkg/store/store.go [hunk 1]
```diff
@@ -160,6 +161,26 @@ type Store struct {
...
+	readIdx := s.raft.LastIndex()
+	// check no leadership transfer
+	if s.raft.VerifyLeader().Error() != nil {
+		return ErrNotLeader
+	}
+	// wait util apply to readIndex
+	done := make(chan struct{})
+	go func() {
+		for s.raft.AppliedIndex() < readIdx {
+			time.Sleep(s.HeartbeatTimeout / 10) // TODO: find a better way
```
Comment (nodece) [inline pkg/store/store.go:175]: I think we can use the https://github.com/cenkalti/backoff here.

**Comment by hsluoyz** [comment]
@noneback don't make WIP PR. First work in your local branch and then create the final PR

**Comment by nodece** [comment]
I need to spend some time for review this PR.

**Comment by noneback** [comment]
> nodece

Thx for your patience！ You can review it at your convenience.

**Comment by WenyXu** [comment]
Well done, But I consider we still need to add tests for this PR. Before adding tests, we need to decouple the main function's parsing flags(the relevant issue #58 ), which makes integration testing more manageable. I will handle this issue later this week. If you are willing, you can make a decoupling of the parsing flags PR, then add tests for this PR.

File: pkg/store/config.go [hunk 2]
```diff
@@ -31,4 +33,5 @@ type StoreConfig struct {
 	AuthType auth.AuthType
 	*auth.CredentialsStore
 	AdvAddr string
+	TcpLn   net.Listener
```
Comment (nodece) [inline pkg/store/config.go:36]: I don't think we need to add this here, it is unnecessary.

File: pkg/store/store.go [hunk 3]
```diff
@@ -160,6 +168,100 @@ type Store struct {
...
+	return readIndex, true
+}
+
+// queryLeaderForReadIndex query read index from leader
+func (s *Store) queryLeaderForReadIndex() (uint64, bool) {
+	if s.State() == Leader {
+		return s.lastCommittedIndex()
+	}
+
+	conn, err := net.Dial("tcp", s.LeaderAddr())
```
Comment (nodece) [inline pkg/store/store.go:190]: We can create a client to request the HTTP server or GRPC server.

**Comment by nodece** [comment]
Thanks for your contribution, most of the work is great, I left my comment.

**Comment by noneback** [comment]
Thx. I will fix them all in my spare time.

File: pkg/store/store.go (hunk 3 above)
Comment (noneback) [reply pkg/store/store.go:190]: My original goal is to reuse the port. Is it necessary to use another port to serve?

Comment (WenyXu) [reply pkg/store/store.go:190]: I suppose this is acceptable, and it’s lighter than gRPC and HTTP. For now, in-node communications, I don't think it benefits from using gRPC or HTTP. Maybe in the future, we can refactor this part by using gRPC. cc @noneback

Comment (nodece) [reply pkg/store/store.go:190]: OK, in-node is lightweight.

**Comment by hsluoyz** [comment]
@noneback can you continue pushing this PR? Also resolve the conflict:

image

**Comment by noneback** [comment]
Sure. I will fix conflict and try to add some test

**Pull request closed without merging**